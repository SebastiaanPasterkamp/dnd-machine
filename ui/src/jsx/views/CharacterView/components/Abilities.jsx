import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    flow,
    map,
} from 'lodash/fp';
import MDReactComponent from 'markdown-react-js';
import {sprintf} from 'sprintf-js';

import Panel from '../../../components/Panel.jsx';

const Abilities = function({ abilities }) {
    return (
        <Panel
            key="abilities"
            className="character-view__abilities info"
            header="Abilities"
        >
            <ul>
            {flow(entries, map(([label, data]) => (
                <li key={label}>
                    <strong>{label}</strong>
                    <MDReactComponent
                        text={sprintf(
                            data.description,
                            data
                        )}
                    />
                </li>
            )))(abilities)}
            </ul>
        </Panel>
    );
};

Abilities.propTypes = {
    abilities: PropTypes.objectOf(
        PropTypes.shape({
            description: PropTypes.string.isRequired,
        })
    ),
};

Abilities.defaultProps = {
    abilities: {},
};

export default Abilities;
