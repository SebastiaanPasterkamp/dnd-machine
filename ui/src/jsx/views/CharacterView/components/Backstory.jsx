import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import Panel from '../../../components/Panel.jsx';

const Backstory = function({ backstory }) {
    if (!backstory) {
        return null;
    }

    return (
        <Panel
            key="backstory"
            className="character-view__backstory info"
            header="Backstory"
        >
            <MDReactComponent text={backstory} />
        </Panel>
    );
};

Backstory.propTypes = {
    description: PropTypes.string,
};

Backstory.defaultProps = {
    description: null,
};

export default Backstory;
