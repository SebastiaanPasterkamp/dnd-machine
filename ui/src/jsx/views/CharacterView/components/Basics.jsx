import React from 'react';
import PropTypes from 'prop-types';

import ListDataWrapper from '../../../hocs/ListDataWrapper.jsx';

import Bonus from '../../../components/Bonus';
import DiceNotation from '../../../components/DiceNotation';
import ListLabel from '../../../components/ListLabel';
import Panel from '../../../components/Panel.jsx';
import Reach from '../../../components/Reach';
import WeightLabel from '../../../components/WeightLabel';

const Basics = function({
    armor_class, armor_class_bonus, hit_points, level,
    hit_dice, speed, height, weight, age, initiative_bonus,
    passive_perception, proficiency, spell, statistics,
}) {
    const {
        stat: spell_stat,
        safe_dc: spell_dc,
        attack_modifier: spell_mod,
    } = spell || {};

    return (
        <Panel
           key="base"
           className="character-view__base info"
           header="Base"
        >
            <tbody>
                <tr>
                    <th>Armor Class</th>
                    {armor_class_bonus ? (
                        <td>
                            {armor_class}
                            &nbsp;/&nbsp;
                            <Bonus
                                bonus={armor_class_bonus}
                            />
                        </td>
                    ) : (
                        <td>{armor_class}</td>
                    )}
                </tr>
                <tr>
                    <th>Hit Points</th>
                    <td>{hit_points}</td>
                </tr>
                <tr>
                    <th>Hit Dice</th>
                    <td>
                        <DiceNotation
                            dice_count={level}
                            dice_size={hit_dice}
                        />
                    </td>
                </tr>
                <tr>
                    <th>Speed</th>
                    <td>
                        <Reach distance={speed} />
                    </td>
                </tr>
                <tr>
                    <th>Height</th>
                    <td>
                        <Reach distance={height} />
                    </td>
                </tr>
                <tr>
                    <th>Weight</th>
                    <td><WeightLabel weight={weight} /></td>
                </tr>
                <tr>
                    <th>Age</th>
                    <td>{age ? `${age} years` : null}</td>
                </tr>
                <tr>
                    <th>Initiative Modifier</th>
                    <td>
                        <Bonus bonus={initiative_bonus} />
                    </td>
                </tr>
                <tr>
                    <th>Passive Perception</th>
                    <td>{passive_perception}</td>
                </tr>
                <tr>
                    <th>Proficiency Bonus</th>
                    <td>
                        <Bonus bonus={proficiency} />
                    </td>
                </tr>

                {spell_stat && (<React.Fragment>
                    <tr>
                        <th>Spell Ability</th>
                        <td>
                            <ListLabel
                                items={statistics}
                                value={spell_stat}
                            />
                        </td>
                    </tr>
                    <tr>
                        <th>Spell DC</th>
                        <td>{spell_dc}</td>
                    </tr>
                    <tr>
                        <th>Spell Attack Modifier</th>
                        <td>
                            <Bonus bonus={spell_mod} />
                        </td>
                    </tr>
                </React.Fragment>)}
            </tbody>
        </Panel>
    );
};

Basics.propTypes = {
    armor_class: PropTypes.number,
    armor_class_bonus: PropTypes.number,
    hit_points: PropTypes.number,
    level: PropTypes.number,
    hit_dice: PropTypes.number,
    speed: PropTypes.number,
    height: PropTypes.number,
    weight: PropTypes.number,
    age: PropTypes.number,
    initiative_bonus: PropTypes.number,
    passive_perception: PropTypes.number,
    proficiency: PropTypes.number,
    spell: PropTypes.shape({
        stat: PropTypes.string.isRequired,
        safe_dc: PropTypes.number.isRequired,
        attack_modifier: PropTypes.number.isRequired,
    }),
    statistics: PropTypes.array,
};

Basics.defaultProps = {
    armor_class: 10,
    armor_class_bonus: null,
    hit_points: 1,
    level: 1,
    hit_dice: 4,
    speed: 30,
    height: null,
    weight: null,
    age: null,
    initiative_bonus: 0,
    passive_perception: 10,
    proficiency: 2,
    spell: null,
    statistics: [],
};

export { Basics };

export default ListDataWrapper(
    Basics,
    [
        'statistics',
    ],
    'items'
);
