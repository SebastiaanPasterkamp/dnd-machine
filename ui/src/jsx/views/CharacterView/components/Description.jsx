import React from 'react';
import PropTypes from 'prop-types';

import CharacterLabel from '../../../components/CharacterLabel.jsx';
import CharacterLinks from '../../../components/CharacterLinks.jsx';
import Panel from '../../../components/Panel.jsx';

const Description = function(props) {
    const { id, name } = props;

    return (
        <Panel
            key="description"
            className="character-view__description info"
            header="Description"
        >
            {id !== null && (
                <CharacterLinks
                    className="pull-right"
                    id={id}
                />
            )}

            <h3>{name}</h3>

            <h4>
                <CharacterLabel
                    showName={false}
                    showInfo={true}
                    showProgress={false}
                    character={props}
                />
            </h4>

            <CharacterLabel
                showInfo={false}
                showProgress={true}
                character={props}
            />
        </Panel>
    );
};

Description.propTypes = {
    id: PropTypes.number,
    name: PropTypes.string,
};

Description.defaultProps = {
    id: null,
    name: 'Unknown',
};

export default Description;
