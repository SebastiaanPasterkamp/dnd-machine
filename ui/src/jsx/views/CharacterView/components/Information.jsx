import React from 'react';
import {
    filter,
    join,
} from 'lodash/fp';
import PropTypes from 'prop-types';

import ListDataWrapper from '../../../hocs/ListDataWrapper.jsx';

import ListLabel from '../../../components/ListLabel.jsx';
import Panel from '../../../components/Panel.jsx';
import UserLabel from '../../../components/UserLabel.jsx';

const Information = function({
    'class': _class, race, background, alignment,
    alignments, xp, adventure_checkpoints, user_id,
}) {
    if (!(race && _class && background)) {
        return null;
    }

    return (
        <Panel
            key="info"
            className="character-view__info info"
            header="Info"
        >
            <thead>
                <tr>
                    <th>Class</th>
                    <th>Background</th>
                    <th>Player</th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td>{_class}</td>
                    <td>{background}</td>
                    <td>
                        <UserLabel user_id={user_id} />
                    </td>
                </tr>
            </tbody>
            <thead>
                <tr>
                    <th>Race</th>
                    <th>Alignment</th>
                    <th>{join(' + ', filter(null, [
                        xp ? 'XP' : null,
                        adventure_checkpoints ? 'ACP' : null,
                    ]))}</th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td>{race}</td>
                    <td>
                        <ListLabel
                            items={alignments}
                            value={alignment}
                        />
                    </td>
                    <td>{join(' + ', filter(null, [
                        xp ? `${xp} XP` : null,
                        adventure_checkpoints
                            ? `${adventure_checkpoints} ACP`
                            : null,
                    ]))}</td>
                </tr>
            </tbody>
        </Panel>
    );
};

Information.propTypes = {
    'class': PropTypes.string,
    race: PropTypes.string,
    background: PropTypes.string,
    alignment: PropTypes.string,
    alignments: PropTypes.array,
    xp: PropTypes.number,
    adventure_checkpoints: PropTypes.number,
    user_id: PropTypes.number,
};

Information.defaultProps = {
    'class': '',
    race: '',
    background: '',
    alignment: '',
    alignments: [],
    xp: null,
    adventure_checkpoints: null,
    user_id: null,
};

export default ListDataWrapper(
    Information,
    [
        'alignments',
    ],
    'items'
);
