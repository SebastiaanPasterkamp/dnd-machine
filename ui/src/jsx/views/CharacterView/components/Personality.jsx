import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';
import {sprintf} from 'sprintf-js';

import Panel from '../../../components/Panel.jsx';

const Personality = function({
    traits, ideals, bonds, flaws,
}) {
    if (!( traits || ideals || bonds || flaws )) {
        return null;
    }

    return (
        <Panel
            key="personality"
            className="character-view__personality info"
            header="Personality"
        >
            <strong>Traits</strong>
            <MDReactComponent text={traits} />

            <strong>Ideals</strong>
            <MDReactComponent text={ideals} />

            <strong>Bonds</strong>
            <MDReactComponent text={bonds} />

            <strong>Flaws</strong>
            <MDReactComponent text={flaws} />
        </Panel>
    );
};

Personality.propTypes = {
    traits: PropTypes.string,
    ideals: PropTypes.string,
    bonds: PropTypes.string,
    flaws: PropTypes.string,
};

Personality.defaultProps = {
    traits: '',
    ideals: '',
    bonds: '',
    flaws: '',
};

export default Personality;
