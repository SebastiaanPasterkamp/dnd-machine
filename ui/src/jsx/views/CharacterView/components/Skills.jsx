import React from 'react';
import PropTypes from 'prop-types';
import {
    includes,
    keys,
    map,
} from 'lodash/fp';

import ListDataWrapper from '../../../hocs/ListDataWrapper.jsx';

import Bonus from '../../../components/Bonus.jsx';
import CheckBox from '../../../components/CheckBox.jsx';
import Panel from '../../../components/Panel.jsx';

const Skills = function({
    statistics,
    skills,
    skillBonus,
    proficiencies: {
        skills: prof_skl,
        expertise: prof_exp,
    },
}) {
    if (!keys(skillBonus).length) {
        return null;
    }

    return (
        <Panel
            className="character-view__skills info"
            header="Skills" contentComponent="table"
        >
        {map((stat) => (
            <React.Fragment key={stat.id}>
                <thead>
                    <tr>
                        <th
                            colSpan="2"
                            className="text-align-center"
                        >
                            {stat.name}
                        </th>
                    </tr>
                </thead>

                <tbody>
                {map((skill) => {
                    if (skill.stat !== stat.id) {
                        return null;
                    }

                    return (
                        <tr key={skill.id}>
                            <th>{skill.name}</th>
                            <td>
                                <CheckBox
                                    isChecked={includes(
                                        skill.id,
                                        prof_skl
                                    )}
                                />
                                {prof_exp.length ? (
                                    <CheckBox
                                        isChecked={includes(
                                            skill.id,
                                            prof_exp
                                        )}
                                    />
                                ) : null}
                                <Bonus bonus={skillBonus[skill.id]} />
                            </td>
                        </tr>
                    );
                })(skills)}
                </tbody>
            </React.Fragment>
        ))(statistics)}
        </Panel>
    );
};

Skills.propTypes = {
    statistics: PropTypes.array,
    skills: PropTypes.array,
    skillBonus: PropTypes.objectOf(
        PropTypes.number,
    ),
    proficiencies: PropTypes.shape({
        skills: PropTypes.array,
        expertise: PropTypes.array,
    }),
};

Skills.defaultProps = {
    statistics: [],
    skills: [],
    skillBonus: {},
    proficiencies: {
        skills: [],
        expertise: [],
    },
};

export { Skills };

export default ListDataWrapper(
    Skills,
    [
        'skills',
        'statistics',
    ],
    'items'
);
