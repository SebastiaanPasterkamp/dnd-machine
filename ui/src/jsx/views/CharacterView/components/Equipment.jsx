import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    flow,
    countBy,
    isObject,
    map,
} from 'lodash/fp';

import { ArmorLabel } from '../../../components/ArmorLabel';
import CountPrefix from '../../../components/CountPrefix';
import Panel from '../../../components/Panel';
import { WeaponLabel } from '../../../components/WeaponLabel';

const Equipment = function({
    weapons, weapon_properties, weapon_types,
    armor, armor_types, items,
}) {
    return (
        <Panel
            key="equipment"
            className="character-view__equipment info"
            header="Equipment"
        >
            <strong>Weapons</strong>
            <ul>
            {map((weapon) => (
                <li key={weapon.id}>
                    <WeaponLabel
                        weapon_properties={weapon_properties}
                        weapon_types={weapon_types}
                        weapon_id={weapon.id}
                        weapon={weapon}
                    />
                </li>
            ))(weapons)}
            </ul>

            <strong>Armor</strong>
            <ul>
            {map((_armor) => (
                <li key={_armor.id}>
                    <ArmorLabel
                        armor_types={armor_types}
                        armor_id={_armor.id}
                        armor={_armor}
                    />
                </li>
            ))(armor)}
            </ul>

            <strong>Gear</strong>

            <ul>
            {map((itemset) => {
                const gear = countBy(
                    (item) => {
                        if (isObject(item)) {
                            return ( item.label || item.name );
                        }
                        return item;
                    },
                    itemset
                );

                return flow(entries, map(([item, cnt]) => (
                    <li key={item}>
                        <CountPrefix count={cnt}>
                            {item}
                        </CountPrefix>
                    </li>
                )))(gear);
            })(items)}
            </ul>
        </Panel>
    );
};

Equipment.propTypes = {
    weapons: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
        })
    ),
    weapon_properties: PropTypes.array,
    weapon_types: PropTypes.array,
    armor: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.number.isRequired,
        })
    ),
    armor_types: PropTypes.array,
    items: PropTypes.objectOf(
        PropTypes.array
    ),
};

Equipment.defaultProps = {
    weapons: [],
    weapon_properties: [],
    weapon_types: [],
    armor: [],
    armor_types: [],
    items: {},
};

export default Equipment;
