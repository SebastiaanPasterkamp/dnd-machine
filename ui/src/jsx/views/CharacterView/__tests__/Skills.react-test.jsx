import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { Skills } from '../components/Skills';

describe('Skills', () => {
    const fullProps = {
        statistics: [
            {
                code: "strength",
                label: "Strength",
            },
            {
                code: "wisdom",
                label: "Wisdom",
            },
        ],
        skills: [
            {
                code: "athletics",
                label: "Athletics",
                stat: "strength",
            },
            {
                code: "insight",
                label: "Insight",
                stat: "wisdom",
            },
            {
                code: "survival",
                label: "Survival",
                stat: "wisdom",
            },
        ],
        skillBonus: {
            athletics: 7,
            insight: -1,
            survival: 1,
        },
        proficiencies: {
            skills: [ "athletics", "survival" ],
            expertise: [ "athletics" ],
        },
    };

    it('should render without props', () => {
        const tree = renderer.create(
            <Skills />
        );

        expect(tree).toMatchSnapshot();
    });

    it('should render with full props', () => {
        const setState = jest.fn();

        const tree = renderer.create(
            <Skills {...fullProps} />
        );

        expect(tree).toMatchSnapshot();
    });
});
