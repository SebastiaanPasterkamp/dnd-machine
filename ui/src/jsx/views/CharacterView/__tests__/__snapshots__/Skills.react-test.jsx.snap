// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skills should render with full props 1`] = `
<div
  className="nice-panel character-view__skills info"
  id={undefined}
>
  <div
    className="nice-panel-heading"
  >
    Skills
  </div>
  <table
    className="nice-panel-content nice-table bordered"
  >
    <thead>
      <tr>
        <th
          className="text-align-center"
          colSpan="2"
        >
          Strength
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          Athletics
        </th>
        <td>
          <span
            className="icon fa-check-square-o"
          >
             
          </span>
          <span
            className="icon fa-check-square-o"
          >
             
          </span>
          <div
            className="bonus inline"
          >
            +
            7
          </div>
        </td>
      </tr>
    </tbody>
    <thead>
      <tr>
        <th
          className="text-align-center"
          colSpan="2"
        >
          Wisdom
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          Insight
        </th>
        <td>
          <span
            className="icon fa-square-o"
          >
             
          </span>
          <span
            className="icon fa-square-o"
          >
             
          </span>
          <div
            className="bonus inline"
          >
            -1
          </div>
        </td>
      </tr>
      <tr>
        <th>
          Survival
        </th>
        <td>
          <span
            className="icon fa-check-square-o"
          >
             
          </span>
          <span
            className="icon fa-square-o"
          >
             
          </span>
          <div
            className="bonus inline"
          >
            +
            1
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Skills should render without props 1`] = `null`;
